microbench/src/main/java/io/netty/buffer/CompositeByteBufSequentialBenchmark.java

```diff
@@ -0,0 +1,159 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer;
+
+import io.netty.microbench.util.AbstractMicrobenchmark;
+import io.netty.util.ByteProcessor;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Warmup;
+
+import static io.netty.buffer.Unpooled.EMPTY_BUFFER;
+import static io.netty.buffer.Unpooled.wrappedBuffer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+@Fork(1)
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 4, time = 1, timeUnit = TimeUnit.SECONDS)
+public class CompositeByteBufSequentialBenchmark extends AbstractMicrobenchmark {
+    
+    public enum ByteBufType {
+        SMALL_CHUNKS {
+            @Override
+            ByteBuf newBuffer(int length) {
+                return newBufferSmallChunks(length);
+            }
+        },
+        LARGE_CHUNKS {
+            @Override
+            ByteBuf newBuffer(int length) {
+                return newBufferLargeChunks(length);
+            }
+        };
+        abstract ByteBuf newBuffer(int length);
+    }
+
+    @Param({ "8", "64", "1024", "10240", "102400", "1024000" }) //"1073741824" })
+    public int size;
+
+    @Param
+    public ByteBufType bufferType;
+
+    private ByteBuf buffer;
+
+    @Override
+    protected String[] jvmArgs() {
+        // Ensure we minimize the GC overhead by sizing the heap big enough.
+        return new String[] { "-Xmx8g", "-Xms8g", "-Xmn6g" };
+    }
+
+    @Setup
+    public void setup() {
+        buffer = bufferType.newBuffer(size);
+    }
+
+    @TearDown
+    public void teardown() {
+        buffer.release();
+    }
+
+    @Benchmark
+    public int forEachByte() {
+        buffer.writerIndex(buffer.capacity()).readerIndex(0);
+        ByteProcessor processor = new ByteProcessor() {
+            @Override
+            public boolean process(byte value) throws Exception {
+                return value == 'b'; // false
+            }
+        };
+        buffer.forEachByte(processor);
+        return buffer.forEachByteDesc(processor);
+    }
+
+    @Benchmark
+    public int sequentialWriteAndRead() {
+        buffer.readerIndex(0).writerIndex(0);
+        for (int i = 0, l = buffer.writableBytes(); i < l; i++) {
+            buffer.writeByte('a');
+        }
+        for (int i = 0, l = buffer.readableBytes(); i < l; i++) {
+            if (buffer.readByte() == 'b') {
+                return -1;
+            }
+        }
+        return 1;
+    }
+
+    protected static ByteBuf newBufferSmallChunks(int length) {
```

Comment:
private ?

Reply:
Sure, I hadn't been paying as much attention to detail in the benchmark class